### PR TITLE
Limit max pool size for dynamic parallel execution

### PR DESCRIPTION
**Pull request by mpkorstanje**
## Overview
This is a followup of #3044 for the `dynamic` strategy.

With Java 9+ the `ForkJoinPool` used by JUnit can be configured with a maximum pool size via the `junit.jupiter.execution.parallel.config.dynamic.max-pool-size-factor` property.

While the number of concurrently executing tests may exceed the configured parallelism factor when tests become blocked, it will not exceed the maximum pool size factor in these circumstances.

With the following configuration:

```properties
junit.jupiter.execution.parallel.enabled=true
junit.jupiter.execution.parallel.config.strategy=dynamic
junit.jupiter.execution.parallel.mode.default=concurrent
junit.jupiter.execution.parallel.config.dynamic.factor=0.1667
```

This example will report between 2 and 4 tests running concurrently on a system with 12 available cores while only (12 * 0.1667 =) 2 would be expected.

```java
class ExampleTest {

    private static final AtomicInteger running = new AtomicInteger();

    @BeforeEach
    void increment() {
        System.out.println("Running " + running.incrementAndGet());
    }

    @AfterEach
    void decrement() {
        running.decrementAndGet();
    }

    static IntStream numbers() {
        return IntStream.range(0, 1000);
    }

    @ParameterizedTest
    @MethodSource("numbers")
    void test(int i) throws ExecutionException, InterruptedException {
        Runnable sleep = () -> {
            try {
                Thread.sleep(600);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        };
        ForkJoinPool.commonPool().submit(sleep).get();
    }
}
```

By also configuring the `max-pool-size-factor` we can ensure the concurrently executing test does not exceed the expected 2 .

```properties
junit.jupiter.execution.parallel.config.dynamic.max-pool-size-factor=1
```

Additionally, because the `ForkJoinPool` will by default reject tasks that would exceed the maximum pool size the
`junit.jupiter.execution.parallel.config.dynamic.saturate` property has been added and will default to `true`. There appears to be no reason to ever set this `false` but it is there should someone depend on the old behavior.

Fixes: #3205

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/DefaultParallelExecutionConfigurationStrategy.java [hunk 1]
```diff
@@ -71,8 +71,20 @@ public ParallelExecutionConfiguration createConfiguration(ConfigurationParameter
...
-			return new DefaultParallelExecutionConfiguration(parallelism, parallelism, 256 + parallelism, parallelism,
-				KEEP_ALIVE_SECONDS, null);
+			int maxPoolSize = configurationParameters.get(CONFIG_DYNAMIC_MAX_POOL_SIZE_FACTOR_PROPERTY_NAME,
+				BigDecimal::new).map(maxPoolSizeFactor -> {
+					Preconditions.condition(maxPoolSizeFactor.compareTo(BigDecimal.ONE) >= 0,
+						() -> String.format(
+							"Factor '%s' specified via configuration parameter '%s' must be greater than or equal to 1",
+							factor, CONFIG_DYNAMIC_FACTOR_PROPERTY_NAME));
+					return maxPoolSizeFactor.multiply(BigDecimal.valueOf(parallelism)).intValue();
+				}).orElseGet(() -> 256 + parallelism);
```
Comment (mpkorstanje) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/DefaultParallelExecutionConfigurationStrategy.java:81]: There are two ways to calculate the max pool size:

```
maxPoolSize = dynamic.max-pool-size-factor * dynamic.factor * Runtime.getRuntime().availableProcessors() 
maxPoolSize = dynamic.max-pool-size-factor * Runtime.getRuntime().availableProcessors() 
```

And the choice is a bit arbitrary. But I think reasoning in terms of "threads per degree of parallelism" is a bit easier so I prefer the former.

**Review by marcphilipp — APPROVED** [review]

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 2]
```diff
@@ -2474,7 +2484,7 @@ The following table lists relevant properties for configuring parallel execution
 | ```junit.jupiter.execution.parallel.config.fixed.max-pool-size```
 | Desired maximum pool size of the underlying fork-join pool for the ```fixed```
   configuration strategy
-| a positive integer, must greater than or equal to `junit.jupiter.execution.parallel.config.fixed.parallelism`
+| a positive integer, must be greater than or equal to `junit.jupiter.execution.parallel.config.fixed.parallelism`
 | 256 + the value of `junit.jupiter.execution.parallel.config.fixed.parallelism`
 
 | ```junit.jupiter.execution.parallel.config.fixed.saturate```
```
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:2490]: We should add an entry for the `dynamic` one as well.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 3]
```diff
@@ -2466,6 +2468,14 @@ The following table lists relevant properties for configuring parallel execution
 | a positive decimal number
 | ```1.0```
 
+| ```junit.jupiter.execution.parallel.config.dynamic.max-pool-size.factor```
```
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:2471]: Why are we using three backticks rather than one in this table?

**Review by marcphilipp — CHANGES_REQUESTED** [review]

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc (hunk 3 above)
Comment (sbrannen) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:2471]: I believe there was a formatting error without the triple backticks.

But maybe that's no longer an issue with more recent versions of Asciidoctor.

So feel free to see what happens if you switch to single backticks.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc (hunk 2 above)
Comment (mpkorstanje) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:2490]: That one is was `junit.jupiter.execution.parallel.config.dynamic.factor` and was already documented here: https://github.com/junit-team/junit5/blame/07f2bf349e45f5a63228cb6002ae7bdbd3445ce1/documentation/src/docs/asciidoc/user-guide/writing-tests.adoc#L2463-L2467

The full set is:
```
junit.jupiter.execution.parallel.config.dynamic.factor
junit.jupiter.execution.parallel.config.dynamic.max-pool-size-factor

junit.jupiter.execution.parallel.config.fixed.parallelism
junit.jupiter.execution.parallel.config.fixed.max-pool-size
```

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc (hunk 3 above)
Comment (mpkorstanje) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: > I believe there was a formatting error without the triple backticks.

This is still the case.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc (hunk 2 above)
Comment (mpkorstanje) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:2498]: https://github.com/junit-team/junit5/pull/3206/commits/6849f00963c4f33c991cfd2b9dd3009ebab461d9

**Review by marcphilipp — APPROVED** [review]

**Pull request merged**